src/hotspot/cpu/aarch64/stubGenerator_aarch64.cpp

```diff
@@ -4860,6 +4860,149 @@ class StubGenerator: public StubCodeGenerator {
+      __ ccmp(tmp1h, tmp2h, 0, Assembler::EQ);
+      __ br(__ NE, DIFF);
+
+      __ ldp(tmp1, tmp1h, Address(str1,16));
+      __ ldp(tmp2, tmp2h, Address(str2,16));
+      __ cmp(tmp1, tmp2);
+      __ ccmp(tmp1h, tmp2h, 0, Assembler::EQ);
+      __ br(__ NE, DIFF);
+
+      __ ldp(tmp1, tmp1h, Address(str1, 32));
+      __ ldp(tmp2, tmp2h, Address(str2, 32));
+      __ cmp(tmp1, tmp2);
+      __ ccmp(tmp1h, tmp2h, 0, Assembler::EQ);
+      __ br(__ NE, DIFF);
+
+      __ ldp(tmp1, tmp1h, Address(str1,48));
+      __ ldp(tmp2, tmp2h, Address(str2,48));
+      __ cmp(tmp1, tmp2);
+      __ ccmp(tmp1h, tmp2h, 0, Assembler::EQ);
+      __ br(__ NE, DIFF);
+
+      __ sub(cnt2, cnt2, isLL ? 64 : 32);
+      __ add(str1, str1, 64);
+      __ add(str2, str2, 64);
+      __ subs(rscratch2, cnt2, largeLoopSize);
+      __ br(__ GE, LOOP_COMPARE64);
+      __ cbz(cnt2, LENGTH_DIFF); // no more chars left?
+
+
+    __ subs(rscratch1, cnt2, isLL ? 16 : 8);
+    __ br(__ LE, LESS16);
+    __ align(OptoLoopAlignment);
+
+    __ bind(LOOP_COMPARE16);
+      __ ldp(tmp1, tmp1h, Address(__ post(str1, 16)));
+      __ ldp(tmp2, tmp2h, Address(__ post(str2, 16)));
+
+      __ cmp(tmp1, tmp2);
+      __ ccmp(tmp1h, tmp2h, 0, Assembler::EQ);
+      __ br(__ NE, DIFF);
+      __ sub(cnt2, cnt2, isLL ? 16 : 8);
+      __ subs(rscratch2, cnt2, isLL ? 16 : 8);
+      __ br(__ LT, LESS16);
+
+      __ ldp(tmp1, tmp1h, Address(__ post(str1, 16)));
+      __ ldp(tmp2, tmp2h, Address(__ post(str2, 16)));
+
+      __ cmp(tmp1, tmp2);
+      __ ccmp(tmp1h, tmp2h, 0, Assembler::EQ);
+      __ br(__ NE, DIFF);
+      __ sub(cnt2, cnt2, isLL ? 16 : 8);
+
+      __ subs(rscratch2, cnt2, isLL ? 16 : 8);
+      __ br(__ GE, LOOP_COMPARE16);
+      __ cbz(cnt2, LENGTH_DIFF);
+
+    __ bind(LESS16);
+      // each 8 compare
+      __ subs(cnt2, cnt2, isLL ? 8 : 4);
+      __ br(__ LE, LESS8);
+      __ ldr(tmp1, Address(__ post(str1, 8)));
+      __ ldr(tmp2, Address(__ post(str2, 8)));
+      __ eor(rscratch2, tmp1, tmp2);
+      __ cbnz(rscratch2, CAL_DIFFERENCE);
+      __ sub(cnt2, cnt2, isLL ? 8 : 4);
+
+    __ bind(LESS8);                          // directly load last 8 bytes
+    if(!isLL) {
+       __ add(cnt2, cnt2, cnt2);
+    }
+      __ ldr(tmp1, Address(str1, cnt2));
+      __ ldr(tmp2, Address(str2, cnt2));
+      __ eor(rscratch2, tmp1, tmp2);
+      __ cbz(rscratch2, LENGTH_DIFF);
+      __ b(CAL_DIFFERENCE);
+
+
+    __ bind(DIFF);
+      __ cmp(tmp1, tmp2);
+      __ csel(tmp1, tmp1, tmp1h, Assembler::NE);
+      __ csel(tmp2, tmp2, tmp2h, Assembler::NE);
+      // reuse rscratch2 register for the result of eor instruction
+      __ eor(rscratch2, tmp1, tmp2);
+
+    __ bind(CAL_DIFFERENCE);
+      __ rev(rscratch2, rscratch2);
+      __ clz(rscratch2, rscratch2);
+      __ andr(rscratch2, rscratch2, isLL ? -8 : -16);
+      __ lsrv(tmp1, tmp1, rscratch2);
+      __ lsrv(tmp2, tmp2, rscratch2);
+      if (isLL) {
+          __ uxtbw(tmp1, tmp1);
```

Comment:
Convention is to indent with two spaces but have four here.

Reply:
Thank you for your suggestion. I have fixed the style and add unalign test case in my new commit.